### ISSUE TITLE
binary_streams.hpp:62: virtual void io::BinaryFileSingleStream::reset(): Assertion `stream_.good()' failed

### ISSUE THREAD

**Issue by spock** [ablab/spades#59]
Hi, and thank you for a great assembler!

I seem to have stumbled upon something trivial, but I can't figure out what the real problem is.
I'm trying to assemble 33 GB of paired + unpaired + merged reads (150 files in total), and I get this stack trace during the K-21 assembly:

```
  0:00:00.164    12M / 12M   INFO    General                 (read_converter.hpp        :  84)   Converting reads to binary format for library #25 (takes a while)
  0:00:00.164    12M / 12M   INFO    General                 (read_converter.hpp        :  85)   Converting paired reads
  0:00:00.285    76M / 80M   INFO    General                 (binary_converter.hpp      : 159)   0 reads written
  0:00:00.292    12M / 80M   INFO    General                 (read_converter.hpp        :  94)   Converting single reads
  0:00:00.511   184M / 184M  INFO    General                 (binary_converter.hpp      : 139)   16384 reads processed
  0:00:00.621   196M / 196M  INFO    General                 (binary_converter.hpp      : 139)   32768 reads processed
  0:00:00.853   224M / 224M  INFO    General                 (binary_converter.hpp      : 139)   65536 reads processed
  0:00:01.214   272M / 272M  INFO    General                 (binary_converter.hpp      : 139)   131072 reads processed
  0:00:01.902   372M / 372M  INFO    General                 (binary_converter.hpp      : 139)   262144 reads processed
  0:00:03.265   572M / 572M  INFO    General                 (binary_converter.hpp      : 139)   524288 reads processed
  0:00:06.032   968M / 968M  INFO    General                 (binary_converter.hpp      : 139)   1048576 reads processed
  0:00:11.603     1G / 1G    INFO    General                 (binary_converter.hpp      : 139)   2097152 reads processed
  0:00:25.231     3G / 3G    INFO    General                 (binary_converter.hpp      : 139)   4194304 reads processed
  0:00:39.642     1G / 3G    INFO    General                 (binary_converter.hpp      : 159)   6602328 reads written
=== Stack Trace ===
[0x40c26a]
[0x46b94d]
[0x470568]
[0x470ceb]
[0x69a48d]
[0x541c71]
[0x40a0ee]
[0x402e62]
[0x7b1ef0]
[0x408f2d]
spades: /spades/src/common/io/reads/binary_streams.hpp:62: virtual void io::BinaryFileSingleStream::reset(): Assertion `stream_.good()' failed.


== Error ==  system call for: "['/workspace/SPAdes-3.11.1-Linux/bin/spades', '/G2-spades/K21/configs/config.info']" finished abnormally, err code: -6
```

The library `#25` is probably the one with `.25_0.cor.fastq.gz` suffix - it's a 900 MB file with exactly 6602328 reads (extendedFrags, a result of merging some paired reads with `flash`). The file is readable, `gunzip --test` reports no problems on it.

I'm not sure if this is relevant, but corrected reads come from a `metaspades.py -m 500` run (which could not handle that many libraries), and this failing `spades.py` run is now limited to `-m 250`.

**Comment by asl** [comment]
This smells like some I/O issue - basically SPAdes tried to rewind a file stream, but OS indicated that something bad happened. Does the issue reproduce if you'd restart SPAdes?

> I'm not sure if this is relevant, but corrected reads come from a metaspades.py -m 500 run (which could not handle that many libraries), and this failing spades.py run is now limited to -m 250.
Note that this does not make much sense - ordinary SPAdes is not suitable for metagenomic assemblies.

**Comment by spock** [comment]
Thank you for a quick response!

Summary: this was an open files limit problem.

Yes, it is reproducible - I tried `--continue` maybe five times, it's always the same message/library.

I have just recompressed that one file with slightly different options (`--fast`, `--best`, `--rsyncable` - all of which should not matter other than re-creating the file), and still got this same error. I have also ran `fqtools validate` on that file, it reported "header mismatch" but that does not seem relevant at all. I have also run `fastQValidator` on that file, and it returned 0 errors.

One more thing that I noticed is that `.bin_reads` contains 1046 files, of which 1020 are `.seq`... I have then increased my limit of open files to 2048 from the standard soft limit 1024, and... **it worked** :)

I wonder if it would be reasonably easy and useful in this case to:
- close files that are not being worked with anymore (2048 is a hard limit on our system, and I just reached it on the 51st library :) );
- show the OS error message if a system call fails.

For now, I'll just merge some/all of single-end+unpaired libraries, to decrease their number.

> Note that this does not make much sense - ordinary SPAdes is not suitable for metagenomic assemblies.

Yep, I understand that results might be very, very chimeric :), and I also do run metaspades on the raw data from this dataset. I'm still curious to try spades, esp. after seeing a preprint "Assembling metagenomes, one community at a time".
